task_4/module/ec2/main.tf

```diff
@@ -0,0 +1,66 @@
+#############
+#DATASOURCES#
+#############
+data "aws_ami" "ami" {
+  most_recent = var.ami_param.most_recent
+  name_regex  = var.ami_param.name_regex
+  owners      = var.ami_param.owners
+}
+
+data "aws_vpc" "vpc" {
+  id = var.vpc_id
+}
+
+data "aws_subnet" "subnet" {
+  vpc_id            = data.aws_vpc.vpc.id
+  availability_zone = var.availability_zone
+  filter {
+    name   = "tag:Name"
+    values = [var.subnet_name]
+  }
+}
+
+data "aws_caller_identity" "current" {}
+
+###########
+#VARIABLES#
+###########
+
+###########
+#RESOURCES#
+###########
+
+resource "aws_instance" "instance" {
+  count         = var.instance_count
+  instance_type = var.instance_type
+  ami           = data.aws_ami.ami.id
+  tags = merge(var.tags, {
+    Name          = "${var.name}_${count.index}"
+    Date_creation = timestamp(),
+    OS_type       = data.aws_ami.ami.platform_details
+    Account_id    = data.aws_caller_identity.current.account_id
+  })
+  root_block_device {
+    volume_type = var.root_volume_param.volume_type
+    volume_size = var.root_volume_param.volume_size
+  }
+  vpc_security_group_ids = var.sg_id_list
+  user_data              = file("${path.module}/userdata.sh")
+  key_name               = var.key_pair_name
+}
+
+resource "null_resource" "data_from_instance" {
```

Comment:
For Task4 only. 
Could you please move this extra step to the root module, Let's say it is not a part  of basic EC2 module.
After you move it to root module, please check if it works for all created instances.

Reply:
new ec2 module output generates an array of ip addresses from the module execution
```terraform
output "instances_public_ip" {
  value = [for i in aws_instance.instance : i.public_ip]
}
```

root module uses it to create a local variable with all public ip addresses from all ec2 module executions
```terraform
locals {
  ec2_pub_ip = flatten([for k, v in module.ec2 : v.instances_public_ip])
}
```
and use it to launch null_resource
```terraform
resource "null_resource" "data_from_instance" {
  count = length(local.ec2_pub_ip)
  connection {
...
    host        = local.ec2_pub_ip[count.index]
...
```